runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,644 @@
+
+  -- if we clear the version from current_result, it'll cause the
+  -- next request to be sent and will also pause new highlights
+  -- from being added in on_win until a new result comes from
+  -- the server
+  if state.current_result then
+    state.current_result.version = nil
+  end
+
+  if state.active_request.request_id then
+    local client = vim.lsp.get_client_by_id(client_id)
+    assert(client)
+    client.cancel_request(state.active_request.request_id)
+    state.active_request = {}
+  end
+end
+
+---@private
+function STHighlighter:on_change()
+  self:reset_timer()
+  if self.debounce > 0 then
+    self.timer = vim.defer_fn(function()
+      self:send_request()
+    end, self.debounce)
+  else
+    self:send_request()
+  end
+end
+
+---@private
+function STHighlighter:reset_timer()
+  local timer = self.timer
+  if timer then
+    self.timer = nil
+    if not timer:is_closing() then
+      timer:stop()
+      timer:close()
+    end
+  end
+end
+
+local M = {}
+
+--- Start the semantic token highlighting engine for the given buffer with the
+--- given client. The client must already be attached to the buffer.
+---
+--- NOTE: This is currently called automatically by |vim.lsp.buf_attach_client()|. To
+--- opt-out of semantic highlighting with a server that supports it, you can
+--- delete the semanticTokensProvider table from the {server_capabilities} of
+--- your client in your |LspAttach| callback or your configuration's
+--- `on_attach` callback.
+---
+--- <pre>lua
+---   client.server_capabilities.semanticTokensProvider = nil
+--- </pre>
+---
+---@param bufnr number
+---@param client_id number
+---@param opts (nil|table) Optional keyword arguments
+---  - debounce (number, default: 200): Debounce token requests
+---        to the server by the given number in milliseconds
+function M.start(bufnr, client_id, opts)
+  vim.validate({
+    bufnr = { bufnr, 'n', false },
+    client_id = { client_id, 'n', false },
+  })
+
+  opts = opts or {}
+  assert(
+    (not opts.debounce or type(opts.debounce) == 'number'),
+    'opts.debounce must be a number with the debounce time in milliseconds'
+  )
+
+  local client = vim.lsp.get_client_by_id(client_id)
+  if not client then
+    vim.notify('[LSP] No client with id ' .. client_id, vim.log.levels.ERROR)
+    return
+  end
+
+  if not vim.lsp.buf_is_attached(bufnr, client_id) then
+    vim.notify(
+      '[LSP] Client with id ' .. client_id .. ' not attached to buffer ' .. bufnr,
+      vim.log.levels.WARN
+    )
+    return
+  end
+
+  if not vim.tbl_get(client.server_capabilities, 'semanticTokensProvider', 'full') then
+    vim.notify('[LSP] Server does not support semantic tokens', vim.log.levels.WARN)
+    return
+  end
+
+  local highlighter = STHighlighter.active[bufnr]
+
+  if not highlighter then
+    highlighter = STHighlighter.new(bufnr)
+    highlighter.debounce = opts.debounce or 200
+  else
+    highlighter.debounce = math.max(highlighter.debounce, opts.debounce or 200)
```

Comment:
Instead of taking the highest debounce of all attached clients in a buffer, we could also just use separate timers per client/per buffer (and that way they could be individually timed/configured). It would complicate things slightly since currently `send_request()` aggregates the clients and sends the request to all of them simultaneously.

In practice, this will almost never matter. I am not aware of any actual use-case for having multiple LSP servers provide semantic tokens at the same time.

Reply:
>  I am not aware of any actual use-case for having multiple LSP servers provide semantic tokens at the same time.

When has that ever stopped anyone?

But you're right, having multiple language servers and possibly wanting more control over which one gets to paint the buffer is a convincing argument for making semantic highlighting opt-in. Let's go with your suggestion then (and add how to do this to the `news.txt` entry).

Reply:
I'd still keep it opt-out instead of opt-in because it's non-intrusive and I think we should rather target the common use case instead of the niche use case. It's not ideal if we have to tell everyone to add a line to their LspAttach handler.

Reply:
I don't have strong opinions either way; it's just slightly easier to add the `start()` mapping to `on_attach` or `LspAttached` than to edit the client capabilities for individual servers.

But if we keep it opt-out, do we want to expose (documented) the `start()` function at all? It's not really needed then, and keeping it private makes it easier to change things around later on (e.g., if that becomes necessary when adding modifiers into the mix).

Just some thoughts; to be clear, @mfussenegger has the last word here.

Reply:
I'll leave everything as-is, for now, then, though this comment specifically was about the debounce timer being aggregated for all attached clients, instead of separated out :)

I'd like to leave start()/stop(). I think it's invaluable to be able to disable it (even if it's enabled for you automatically) for debugging purposes. Creating a simple toggler in my config was invaluable to be able to see the differences the semantic token highlighting was making, and I think there are probably other use cases people could come up with for being able to enable/disable the highlighting in a buffer without wholesale disabling the feature entirely.